packages/miniflare/src/workers/core/entry.worker.ts

```diff
@@ -33,10 +34,30 @@ function getUserRequest(
 	request: Request<unknown, IncomingRequestCfProperties>,
 	env: Env
 ) {
+	// The original URL is a header that is passed by Miniflare to the user worker
+	// in case the request was rewritten on its way to the user worker.
 	const originalUrl = request.headers.get(CoreHeaders.ORIGINAL_URL);
-	const upstreamUrl = env[CoreBindings.TEXT_UPSTREAM_URL];
-	let upstreamHost: string|undefined;
 	let url = new URL(originalUrl ?? request.url);
+
+	// The `upstreamHost` is used to override the `Host` header on the request being handled.
+	let upstreamHost: string | undefined;
+
+	// If the request is signed by an upstream proxy then we can use the one from the ORIGINAL_URL.
+	// The signature is required to prevent a malicious user being able to change the host header without permission.
+	const proxySignature = request.headers.get(CoreHeaders.PROXY_SIGNATURE);
+	if (proxySignature) {
```

Comment:
Does this circumvent the protection by just not setting a signature? Should the request be blocked if no signature was provided? So, instead of checking if it matches `env[CoreBindings.TEXT_PROXY_SIGNATURE]` only if it's provided, it should check it matches no matter what

Reply:
nit: I don't think this can be called a signature – a signature uses the content as input but this looks like it's just a shared secret

Reply:
> Does this circumvent the protection by just not setting a signature? Should the request be blocked if no signature was provided? So, instead of checking if it matches `env[CoreBindings.TEXT_PROXY_SIGNATURE]` only if it's provided, it should check it matches no matter what

What we are stopping here is updating the sensitive Host header if the request does not come from the Proxy. Requiring the signature is always present in the header would be a tighter security requirement, yes. But I am not sure we need that extra level of security.

> nit: I don't think this can be called a signature – a signature uses the content as input but this looks like it's just a shared secret

You are probably right here. I called it a signature because we are "signing" the request with a shared secret in a request header to prove who sent it. But I can see how that can be confusing given the standard cryptographic definition of a digital signature. I'll update to use "Shared Secret" terminology instead.